onadata/libs/utils/project_utils.py

```
@@ -49,29 +58,45 @@ def set_project_perms_to_xform(xform, project):
                 role.add(user, xform)
 
 
+# pylint: disable=invalid-name
 @task(bind=True, max_retries=3)
 def set_project_perms_to_xform_async(self, xform_id, project_id):
-
+    """
+    Apply project permissions for ``project_id`` to a form ``xform_id`` task.
+    """
     def _set_project_perms():
         try:
             xform = XForm.objects.get(id=xform_id)
             project = Project.objects.get(id=project_id)
         except (Project.DoesNotExist, XForm.DoesNotExist) as e:
             msg = '%s: Setting project %d permissions to form %d failed.' % (
-                    type(e), project_id, xform_id)
-            report_exception(msg, e, sys.exc_info())
-            self.retry(countdown=60, exc=e)
+                type(e), project_id, xform_id)
+            # make a report only on the 3rd try.
+            if self.request.retries > 2:
+                report_exception(msg, e, sys.exc_info())
+            self.retry(countdown=60 * self.request.retries, exc=e)
         else:
             set_project_perms_to_xform(xform, project)
 
     try:
-        if len(getattr(settings, 'SLAVE_DATABASES', [])):
+        if getattr(settings, 'SLAVE_DATABASES', []):
             from multidb.pinning import use_master
             with use_master:
                 _set_project_perms()
         else:
             _set_project_perms()
-    except Exception as e:
+    except (Project.DoesNotExist, XForm.DoesNotExist) as e:
+        # make a report only on the 3rd try.
+        if self.request.retries > 2:
+            msg = '%s: Setting project %d permissions to form %d failed.' % (
+                type(e), project_id, xform_id)
+            report_exception(msg, e, sys.exc_info())
+        # let's retry if the record may still not be available in read replica.
+        self.retry(countdown=60 * self.request.retries)
+    except IntegrityError:
+        # nothing to do, fail silently.
+        pass
```

why is it OK to fail silently here?